### PR TITLE
Harden WP eCommerce against APC cache deserialization errors

### PR DESCRIPTION
**Pull request by JeffPyeBrook**
Harden countries class against bad cache based on issues noted on sites using GoDaddy APC caching reported here:

https://wordpress.org/support/topic/fatal-error-wpsc_countries/page/2?replies=38#post-6810134

### CODE REVIEW

File: wpsc-includes/wpsc-countries.class.php [hunk 1]
```diff
@@ -1071,23 +1071,36 @@ public static function save() {
...
-		if ( is_array( $data ) ) {
-			foreach ( $data as $variable => $value ) {
-				if ( property_exists( $this, $variable ) ) {
+		$transient_is_valid = true;
+		foreach( self::$_maps_to_save_with_core_class as $map_name => $should_be_saved ) {
+			if ( $should_be_saved ) {
+				if ( ! is_a( $data[ $map_name ], 'WPSC_Data_Map' ) ) {
+					$transient_is_valid = false;
+					delete_transient( self::transient_name() );
+					error_log( __FUNCTION__ . ' transient data corrupted' );
```
Comment (JustinSainton) [inline wpsc-includes/wpsc-countries.class.php]: Let's remove this.

Comment (JeffPyeBrook) [inline wpsc-includes/wpsc-countries.class.php]: remove the error log? or the whole check?

**Comment by JeffPyeBrook** [comment]
@JustinSainton can you give this PR a read and a test

**Comment by JustinSainton** [comment]
@JeffPyeBrook Would like to see 273acd14bd18498fd30859b996cd9057cf0b410d mostly reverted, aside from the function definitions.  Would like to keep this PR as tightly scope and low-touch as possible.  We're dealing with an edge case here, and I'd like the PR to solve the reported issue.  Trying to solve issues we're not yet having can only lead to pain and suffering.

**Comment by JustinSainton** [comment]
@JeffPyeBrook Can you update the PR to reflect my last comment?  I'd like to see this land in 4.0.  Thanks!

File: wpsc-includes/misc.functions.php [hunk 2]
```diff
@@ -988,3 +988,79 @@ function _wpsc_remove_erroneous_files() {
...
+
+			// if there was a transient, but it could not be decoded, we delete the transient to get back
+			// to a working state
+			if ( false === $value ) {
+				_wpsc_delete_transient( $transient );
+			}
+		}
+	}
+
+	return $value;
```
Comment (JustinSainton) [inline wpsc-includes/misc.functions.php]: Value is not defined for all paths.  `$encoded_value`, maybe?

File: wpsc-includes/misc.functions.php [hunk 3]
```diff
@@ -988,3 +988,79 @@ function _wpsc_remove_erroneous_files() {
...
+ */
+function _wpsc_get_transient( $transient )  {
+	$encoded_value = get_transient( $transient );
+	if ( false !== $encoded_value ) {
+		if ( ! empty( $encoded_value ) && is_string( $encoded_value ) ) {
+			$serialized_value = @base64_decode( $encoded_value );
+			if ( is_string( $serialized_value ) ) {
+				$value = unserialize( $serialized_value );
+				if ( ! $value ) {
+					xdebug_break();
```
Comment (JustinSainton) [inline wpsc-includes/misc.functions.php]: hmmm....?

**Comment by JustinSainton** [comment]
haha, what is going on here?

**Comment by JeffPyeBrook** [comment]
change localized in this branch JeffPyeBrook:harden-countries-class-against-bad-cache-redux., this PR is no longer

**Pull request closed without merging**